File: app/src/main/java/com/badradio/nz/services/RadioManager.java
package com.badradio.nz.services;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;

import com.badradio.nz.utilities.Log;
import com.badradio.nz.utilities.Tools;

public class RadioManager {

    private static RadioManager instance = null;
    private static RadioService service;
    private boolean serviceBound;

    private RadioManager() {
        serviceBound = false;
    }

    public static RadioManager with() {
        if (instance == null)
            instance = new RadioManager();
        return instance;
    }

    public static RadioService getService() {
        return service;
    }

    public void playOrPause(String streamUrl) {
        if (streamUrl == null)
            service.stop();
        else
            service.playOrPause(streamUrl);
    }

    public void stopServices() {

        service.stopForeground(true);
        service.stop();
    }

    public boolean isPlaying() {
        return service.isPlaying();
    }

    public void bind(Context context) {
        //Perhaps also catch a LeakedServiceConnection, and if caught: then call unbind and then try to bind again
        if (!serviceBound) {
            Intent intent = new Intent(context, RadioService.class);
            context.startService(intent);
            boolean bound = context.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
            if (service != null)
                Tools.onEvent(service.getStatus());
        }
    }

    public void unbind(Context context) {
        if (serviceBound) {
            try {
                service.stop();
                context.unbindService(serviceConnection);
                context.stopService(new Intent(context, RadioService.class));
                serviceBound = false;
            } catch (IllegalArgumentException e) {
                Log.printStackTrace(e);
            }
        }
    }

    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName arg0, IBinder binder) {
            service = ((RadioService.LocalBinder) binder).getService();
            serviceBound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {
            serviceBound = false;
        }
    };

}

File: app/src/main/java/com/badradio/nz/services/RadioManager.kt
package com.badradio.nz.services

import com.badradio.nz.utilities.Tools.onEvent
import com.badradio.nz.utilities.Log.printStackTrace
import android.os.IBinder
import com.badradio.nz.services.RadioService.LocalBinder
import android.content.*
import java.lang.IllegalArgumentException

class RadioManager private constructor() {
    private var serviceBound = false
    fun playOrPause(streamUrl: String?) {
        if (streamUrl == null) service!!.stop() else service!!.playOrPause(streamUrl)
    }

    fun stopServices() {
        service!!.stopForeground(true)
        service!!.stop()
    }

    val isPlaying: Boolean
        get() = service!!.isPlaying

    fun bind(context: Context) {
        //Perhaps also catch a LeakedServiceConnection, and if caught: then call unbind and then try to bind again
        if (!serviceBound) {
            val intent = Intent(context, RadioService::class.java)
            context.startService(intent)
            val bound = context.bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
            if (service != null) onEvent(service!!.status!!)
        }
    }

    fun unbind(context: Context) {
        if (serviceBound) {
            try {
                service!!.stop()
                context.unbindService(serviceConnection)
                context.stopService(Intent(context, RadioService::class.java))
                serviceBound = false
            } catch (e: IllegalArgumentException) {
                printStackTrace(e)
            }
        }
    }

    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(arg0: ComponentName, binder: IBinder) {
            service = (binder as LocalBinder).service
            serviceBound = true
        }

        override fun onServiceDisconnected(arg0: ComponentName) {
            serviceBound = false
        }
    }

    companion object {
        private var instance: RadioManager? = null
        var service: RadioService? = null
            private set

        fun with(): RadioManager? {
            if (instance == null) instance = RadioManager()
            return instance
        }
    }
}